Hw06 atm emulator

FlameNur opened the pull request.

FlameNur committed:
Предварительный вариант (вер. 1)

FlameNur committed:
Предварительный вариант (вер. 1)

FlameNur committed:
Окончательный вариант (вер. 1)

FlameNur committed:
Окончательный вариант (вер. 1)

FlameNur committed:
Окончательный вариант (вер. 1)

FlameNur committed:
Версия 2.0

FlameNur committed:
Версия 2.1

FlameNur committed:
Версия 2.2

FlameNur committed:
Версия 2.3

FlameNur committed:
Версия 2.4 (недоработан в части WithdrawStrategy)

FlameNur committed:
Версия 2.5 (недоработан в части WithdrawStrategy)

FlameNur committed:
Версия 2.6 (недоработан в части проверки количества банкнот для выдачи)

FlameNur committed:
Версия 3.0

Granomir commented on hw06-ATM-Emulator/src/main/java/com/nuramov/hw06_ATM_Emulator/AtmExample.java:
```diff
@@ -0,0 +1,101 @@
+    private final Map<Integer, Integer> banknoteCells = new TreeMap<>();
+
+    {
+        banknoteCells.put(Rub.RUB_5000.getValue(), 0);
```
в качестве ключей тоже можно использовать enum

Granomir commented on hw06-ATM-Emulator/src/main/java/com/nuramov/hw06_ATM_Emulator/AtmExample.java:
```diff
@@ -0,0 +1,101 @@
+
+        // Определяем количество выдаваемых банкнот
+        // Количество банкнот может меняться в заисимости от выбранной стратегии в классе Main
+        paymentCells = withdrawStrategy.payment(money, paymentCells, banknoteCells);
```
если ты хочешь чтобы стратегия с самими исходными ячейками ничё не делала, а сообщала только сколько и каких банкнот надо выдать, то ты можешь в неё передавать просто копию мапы, т.е. тут создать новую мапу и скопировать в неё содержимое banknoteCells, вот тогда это будут реально разные объекты. Но вот перепроверять что там насчитала стратегия в методе банкомата смысла нет, это ответственность стратегии. Банкомат должен только воспользоваться результатами - предполагается что стратегия правильно выполняет свою работу, проверяется она развечто тестами

Granomir commented on hw06-ATM-Emulator/src/main/java/com/nuramov/hw06_ATM_Emulator/AtmExample.java:
```diff
@@ -0,0 +1,101 @@
+                // Банкноты сначала снимаются из banknoteCells, а если их недостаточно для выдачи,
+                // обратно возвращаются в banknoteCells
+                banknoteCells.put(m.getKey(), m.getValue());
+                System.out.println("Вы хотите снять: " + money);
```
я так понимаю это нужно вынести из цикла

Granomir commented on hw06-ATM-Emulator/src/main/java/com/nuramov/hw06_ATM_Emulator/AtmExample.java:
```diff
@@ -0,0 +1,101 @@
+    // Проверяем баланс Atm.
+    // Если денег недостаточно, возвращаем true и завершаем операцию
+    private boolean checkBalance(int money) {
+        boolean b = false;
```
тут как-то странно выглядит что возвращаем false в случае успешного результата проверки. Как то интуитивно хотелось бы получить true в ответ на проверку что деньги есть. Либо надо тогда назвать метод по другому, типа isNotEnoughMoney(int requestedMoney)

FlameNur committed:
Версия 3.1

FlameNur committed:
Версия 3.2
1 В качестве ключей в banknoteCells использование enum не совсем удобно, приходится в каждом методе, где использован banknoteCells, Rub конвертировать в Integer и обратно, поэтому оставил как есть.
2 Выполнил. Проверка осуществляется в стратегии.
3 Выполнил.
4 Выполнил.

Осталось реализовать проверки через junit

FlameNur committed:
Версия 3.3

FlameNur committed:
Версия 4.0

FlameNur committed:
Версия 4.1

FlameNur committed:
Версия 4.2
